src/operator/nn/pooling.cc

```diff
@@ -421,11 +463,16 @@ NNVM_REGISTER_OP(_backward_Pooling)
 .set_attr<nnvm::FInplaceOption>(
     "FInplaceOption",
     [](const NodeAttrs &attrs) {
-#if MXNET_USE_CUDNN == 1
-  return std::vector<std::pair<int, int> >();
-#else
-  return std::vector<std::pair<int, int> >{{1, 0}};
+#if MXNET_USE_MKLDNN == 1 && MXNET_USE_CUDA == 0 && MXNET_USE_CUDNN == 0
```

Comment:
So unfortunately, all enabled backend operator implementations must live with the same FInPlaceOptions.  I voiced concern about this restriction to @eric-haibin-lin and he graciously put some thought into: https://github.com/apache/incubator-mxnet/issues/13598. The issue has seen no activity in a month, but I'd like to see it reawakened.

Until then, the code you highlight enforces that the most conservatinve "least common denominator" FInPlaceOptions (i.e. none) are in effect for the enabled Pooling implementations. Since an operator cannot rely on FInPlace actually occurring, there should be no functional regression. And perf-wise, a user must understand that at this point enabling multiple back-ends may have some performance consequences due to the FInPlaceOption implementation.

Reply:
Sorry for any confusion.  Perhaps my comments in the code for FInplaceOption speak to the issue more directly:
```
  // Since this routine is not provided the cpu/gpu context info, only in the case
  // where CUDA and CUDNN implementations are not available can we be sure the MKLDNN
  // implementation will be employed.  The MKLDNN FInplaceOptions are not compatible
  // with the other (i.e. cpu, cuda and cudnn) implementations.
```

Reply:
Thanks for the explanation. I understand the current limitation in here. 
My concern is the performance regression in the CPU side when _mxnet-cuXXmkl_ binary is used which actually is the widely used binary like sockeye, gluonNLP, gluonCV, etc. 
Before we resolved issue #13598 (btw, we're working on this now）, I think we need to keep this line **AS-IS**.

Reply:
mxnet-cuXX-mkl pip binary statically links cudnn and will not be affected. @szha correct me if I am wrong. 

The potentially affected cases are:
- build mxnet with mkldnn and cuda, but without cudnn 
- build mxnet with neither mkldnn nor cuda.

As this operator is used in backward for training, I see the two cases above quite rare. Therefore, I do not think this is a blocking issue for this PR. 

Nonetheless, I suggest @DickJC123 create a github issue with this change so that people are aware and can be tracked later, and preferably run some performance testing to have a rough idea how big an impact it is making for the above two cases.

Let's continue discussions on  #13598 to avoid such problems in the future.

Reply:
I originally set out to create a test case (and separate issue) to demonstrate the need for the  restrictive #if clause I put around the MKLDNN FInplaceOption.  I was unable to do so and in the process I came to realize that I misunderstood the meaning behind the {1,0} option.  In fact, it allows the framework to share the backward '1' input (the unused workspace gradient) with the backward '0' output (the pooling input gradient) where the additional MKLDNN workspace is present.  This sharing is compatible with CUDA and cuDNN Pooling implementations, which do not use a workspace or its gradient!  I'm therefore reverting back to a simple code guard around the MKLDNN FinplaceOption- one whose logic mirrors the other MKLDNN/no-MKLDNN switches of behavior within nn/pooling.cc.  No new issue is required.  With my last commit I've also expanded the pooling tests to cover cases that have the inplace tensor sharing, which requires the pooling input to be the same size as the output (and also then the workspace).

Reply:
Good to know MKLDNN impl is not affected